### PR TITLE
added mssign command extension for PRSS key isolation and p7b support

### PR DESCRIPTION
**Pull request by zhili1208**
1. added mssign command extension for nuget.exe
2. added native sign code path for key isolation.

### CODE REVIEW

File: src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+using NuGet.CommandLine;
+using NuGet.Commands;
+using NuGet.Common;
+using NuGet.Packaging.Signing;
+using System.Security.Cryptography;
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs]: sort

File: src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+        private void WarnIfNoTimestamper(ILogger logger)
+        {
+            if (string.IsNullOrEmpty(Timestamper))
+            {
+                logger.Log(LogMessage.CreateWarning(NuGetLogCode.NU3521, NuGetMSSignCommand.MSSignCommandNoTimestamperWarning));
+            }
+        }
+
+        private void ValidateOutputDirectory()
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs]: This should be EnsureOutputDirectory or something related.

File: src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        {
+            if (!string.IsNullOrEmpty(OutputDirectory) &&
+                            !Directory.Exists(OutputDirectory))
+            {
+                Directory.CreateDirectory(OutputDirectory);
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs]: The exists check isn't needed, CreateDirectory handles this

File: src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+            }
+        }
+
+        private void ValidateCertificateInputs()
+        {
+            if (string.IsNullOrEmpty(CertificateFile)
+                && Path.GetExtension(CertificateFile).Equals(".p7b", StringComparison.InvariantCultureIgnoreCase))
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs]: :bug: this is an impossible condition, the value cannot be both null and have that extension (unless the property getter returns something else the second time!)

Comment (emgarten) [reply src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs]: I would use `CertificateFile?.EndsWith(".p7b", StringComparer.OrdinalIgnoreCase)`

Comment (emgarten) [reply src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs]: OrdinalIgnoreCase would be better than InvariantCultureIgnoreCase here unless I'm missing something

File: src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+            {
+                // Throw if user gave no certificate input
+                throw new ArgumentException(NuGetMSSignCommand.MSSignCommandNoValidCertificateFileException);
+            }
+
+            if (string.IsNullOrEmpty(CSPName))
+            {
+                throw new ArgumentException(string.Format(CultureInfo.InvariantCulture,
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs]: Use current culture

File: src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+            }
+        }
+
+        private Common.HashAlgorithmName ValidateAndParseHashAlgorithm(string value, string name, SigningSpecifications spec)
+        {
+            var hashAlgorithm = Common.HashAlgorithmName.SHA256;
+
+            if (!string.IsNullOrEmpty(value))
+            {
+                if (!spec.AllowedHashAlgorithms.Contains(value, StringComparer.InvariantCultureIgnoreCase) ||
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs]: Use `CryptoHashUtility` if possible, or add a common helper to parse this. Culture seems incorrect here

**Review by emgarten — CHANGES_REQUESTED** [review]
Clean up sign command code

File: src/NuGet.Clients/NuGet.CommandLine/Commands/SignCommand.cs [hunk 7]
```diff
@@ -117,17 +109,21 @@ public SignArgs GetSignArgs()
...
         private HashAlgorithmName ValidateAndParseHashAlgorithm(string value, string name, SigningSpecifications spec)
         {
-            var hashAlgorithm = HashAlgorithmName.SHA256;
+            var hashAlgorithm = Common.HashAlgorithmName.SHA256;
 
             if (!string.IsNullOrEmpty(value))
             {
-                if (!spec.AllowedHashAlgorithms.Contains(value, StringComparer.InvariantCultureIgnoreCase) ||
-                    !Enum.TryParse(value, ignoreCase: true, result: out hashAlgorithm))
+                if (!spec.AllowedHashAlgorithms.Contains(value, StringComparer.InvariantCultureIgnoreCase))
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/SignCommand.cs]: Use ordinal ignore case compare

File: src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+            if (!string.IsNullOrEmpty(OutputDirectory))
+            {
+                Directory.CreateDirectory(OutputDirectory);
+            }
+        }
+
+        private void ValidateCertificateInputs()
+        {
+            if (string.IsNullOrEmpty(CertificateFile)
+                || CertificateFile.EndsWith(".p7b", StringComparison.OrdinalIgnoreCase))
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs]: I think you meant to negate the second condition.

However, I think you should remove this second condition altogether.  The file will be a PKCS #7-formatted file containing one or more certificates.  The file extension is unimportant.  .NET figures out the [file format](https://referencesource.microsoft.com/#mscorlib/system/security/cryptography/x509certificates/x509certificate.cs,784) without considering the file extension.

File: src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+
+            return cngkey;
+        }
+
+        private X509Certificate2 GetCertificate()
+        {
+            var certCollection = new X509Certificate2Collection();
+            certCollection.Import(CertificateFile);
+
+            var matchingCertCollection = certCollection.Find(X509FindType.FindByThumbprint, CertificateFingerprint, false);
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.MSSigning.Extensions/MSSignCommand.cs]: Add a named parameter to `false` to explain what it means.

File: src/NuGet.Core/NuGet.Packaging/Signing/Authoring/SignPackageRequest.cs [hunk 10]
```diff
@@ -27,5 +27,12 @@ public class SignPackageRequest
         /// </summary>
         public X509Certificate2 Certificate { get; set; }
 
+#if IS_DESKTOP
+        /// <summary>
+        /// PrivateKey is only used in PRSS sign.
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Authoring/SignPackageRequest.cs]: PrivateKey is only used by mssign command.

File: src/NuGet.Core/NuGet.Packaging/Signing/Authoring/X509SignatureProvider.cs [hunk 11]
```diff
@@ -78,6 +84,32 @@ private Task<Signature> CreateSignature(X509Certificate2 cert, SignatureManifest
...
+            CryptographicAttributeObjectCollection attributes
+            )
+        {
+            var cms = NativeUtilities.NativeSign(
+                signatureManifest.GetBytes(), cert, privateKey, attributes, hashAlgorithm);
+
+            return Task.FromResult(Signature.Load(cms));
+        }
+
+        private CryptographicAttributeObjectCollection GetSignAttributes()
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Authoring/X509SignatureProvider.cs]: This is going to involve a lot of duplicated code.  Is it feasible for this project to add a reference to NuGet.Protocol and call the same method that the sign command will use to get signed attributes?

You could also deduplicate a lot of resource strings that way; reuse sign command strings where possible but create mssign command strings as necessary.

File: src/NuGet.Core/NuGet.Packaging/Signing/Authoring/X509SignatureProvider.cs [hunk 12]
```diff
@@ -78,6 +84,32 @@ private Task<Signature> CreateSignature(X509Certificate2 cert, SignatureManifest
...
+        {
+            var attributes = new CryptographicAttributeObjectCollection();
+
+            var signingTime = new Pkcs9SigningTime();
+            attributes.Add(new CryptographicAttributeObject(
+                    signingTime.Oid,
+                    new AsnEncodedDataCollection(signingTime)));
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Authoring/X509SignatureProvider.cs]: Lines 105-108 can be simplified to:

    attributes.Add(new Pkcs9SigningTime());

File: src/NuGet.Core/NuGet.Packaging/Signing/Authoring/X509SignatureProvider.cs [hunk 13]
```diff
@@ -59,16 +59,22 @@ public Task<Signature> CreateSignatureAsync(SignPackageRequest request, Signatur
         }
 
 #if IS_DESKTOP
-        private Task<Signature> CreateSignature(X509Certificate2 cert, SignatureManifest signatureManifest)
+        private Task<Signature> CreateSignature(SignPackageRequest request, SignatureManifest signatureManifest)
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Authoring/X509SignatureProvider.cs]: It's a private method, so I don't see why it needs to be async.

File: src/NuGet.Core/NuGet.Packaging/Signing/Authoring/X509SignatureProvider.cs [hunk 14]
```diff
@@ -78,6 +84,32 @@ private Task<Signature> CreateSignature(X509Certificate2 cert, SignatureManifest
             return Task.FromResult(Signature.Load(cms));
         }
 
+        private Task<Signature> CreateSignature(
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Authoring/X509SignatureProvider.cs]: Does this need to be async?

File: src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+                Marshal.ThrowExceptionForHR(Marshal.GetHRForLastWin32Error());
+            }
+        }
+#if IS_DESKTOP
+        internal static SignedCms NativeSign(byte[] data, X509Certificate2 certificate, CngKey privateKey, CryptographicAttributeObjectCollection attributes, Common.HashAlgorithmName hashAlgorithm)
+        {
+            using (privateKey)
+            using (var hb = new HeapBlockRetainer())
+            {
+                var chain = new X509Chain();
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs]: Wrap in a `using` statement.

File: src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+
+                // Skipping a bunch of chain building options we should normally set otherwise
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs]: Let's do this before you merge.

File: src/NuGet.Core/NuGet.Packaging/Signing/Authoring/SignPackageRequest.cs [hunk 17]
```diff
@@ -27,5 +27,12 @@ public class SignPackageRequest
...
+        /// </summary>
+        public System.Security.Cryptography.CngKey PrivateKey { get; set; }
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Authoring/SignPackageRequest.cs]: `SignPackageRequest` should implement `IDisposable` and dispose of this member and `Certificate`.  Instances of `SignPackageRequest` should be wrapped in a `using` statement.

File: src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+        {
+            if (!result)
+            {
+                Marshal.ThrowExceptionForHR(Marshal.GetHRForLastWin32Error());
+            }
+        }
+#if IS_DESKTOP
+        internal static SignedCms NativeSign(byte[] data, X509Certificate2 certificate, CngKey privateKey, CryptographicAttributeObjectCollection attributes, Common.HashAlgorithmName hashAlgorithm)
+        {
+            using (privateKey)
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs]: Shouldn't dispose here if a caller wraps `SignPackageRequest` in a `using` statement.

File: src/NuGet.Core/NuGet.Packaging/Signing/Authoring/X509SignatureProvider.cs (hunk 11 above)
Comment (emgarten) [reply src/NuGet.Core/NuGet.Packaging/Signing/Authoring/X509SignatureProvider.cs]: Protocol depends on Packaging, the helper utility should move here and also be used by NuGet.Protocol

File: src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeMethods.cs [hunk 19]
```diff
@@ -209,4 +242,78 @@ internal enum CMSG_GETPARAM_TYPE : uint
...
+        internal int cCertEncoded;
+        internal IntPtr rgCertEncoded;
+        internal int cCrlEncoded;
+        internal IntPtr rgCrlEncoded;
+        internal int cAttrCertEncoded;
+        internal IntPtr rgAttrCertEncoded;
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    public struct CERT_CONTEXT
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeMethods.cs]: Can you make this and any other types in this file `internal`?

File: src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using NuGet.Common;
+#endif
+
+namespace NuGet.Packaging.Signing
+{
+    internal static class NativeUtilities
+    {
+        internal static void SafeFree(IntPtr ptr)
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs:18]: Do you still need this with your `HeapBlockRetainer`?  If no, please remove.

File: src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+            }
+        }
+#if IS_DESKTOP
+        internal static SignedCms NativeSign(byte[] data, X509Certificate2 certificate, CngKey privateKey, CryptographicAttributeObjectCollection attributes, Common.HashAlgorithmName hashAlgorithm)
+        {
+            using (var hb = new HeapBlockRetainer())
+            using (var chain = new X509Chain())
+            {
+                // Skipping a bunch of chain building options we should normally set otherwise
+                chain.Build(certificate);
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs:40]: Some options need to be set, some additional certs need to be added, and the return result must be checked.

File: src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+            CryptographicAttributeObjectCollection attributes,
+            Common.HashAlgorithmName hashAlgorithm,
+            HeapBlockRetainer hb)
+        {
+            var signerInfo = new CMSG_SIGNER_ENCODE_INFO();
+            signerInfo.cbSize = (uint)Marshal.SizeOf(signerInfo);
+            signerInfo.pCertInfo = Marshal.PtrToStructure<CERT_CONTEXT>(certificate.Handle).pCertInfo;
+            signerInfo.hCryptProvOrhNCryptKey = privateKey.Handle.DangerousGetHandle();
+            // DwKeySpec is not used when hCryptProvOrhNCryptKey is used.
+            signerInfo.dwKeySpec = 1;
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs]: Just remove lines 130-131.

File: src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+            signerInfo.HashAlgorithm.pszObjId = hashAlgorithm.ConvertToOidString();
+
+            if (attributes.Count != 0)
+            {
+                signerInfo.cAuthAttr = attributes.Count;
+
+                checked
+                {
+                    var attributeSize = Marshal.SizeOf(typeof(CRYPT_ATTRIBUTE));
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs]: Can't you use `Marshal.SizeOf<T>()`?  I think there's a compiler warning for using the old `Marshal.SizeOf(Type)`.

File: src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+                    var blobSize = Marshal.SizeOf(typeof(CRYPT_INTEGER_BLOB_INTPTR));
+                    var attributesArray = (CRYPT_ATTRIBUTE*)hb.Alloc(attributeSize * attributes.Count);
+                    var attributeIndex = attributesArray;
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs]: Perhaps `attributeIndex` ---> `currentAttribute`?

File: src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs [hunk 25]
```diff
@@ -0,0 +1,178 @@
...
+                                var rawData = hb.Alloc(value.RawData.Length);
+
+                                // Assign data to datablob
+                                dataBlob->cbData = (uint)attrData.Length;
+                                dataBlob->pbData = rawData;
+
+                                Marshal.Copy(attrData, 0, rawData, attrData.Length);
+                            }
+                        }
+                        // Update attribute index
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs]: The comment doesn't add value over the following statement.  Can remove it.

**Review by dtivel — APPROVED** [review]

**Pull request merged**